### PR TITLE
Proxied sites websocket implementation

### PR DESCRIPTION
**Pull request by atavism**
This contains the following changes to the proxied sites functionality:
- Use a websocket
- Pass add/delete deltas instead of complete proxied sites list between requests
- UI screen changes depending on backend availability
- websocket automatically reconnects on disconnect (max re-connect attempt is set to 20)

Still QAing this.

### CODE REVIEW

**Comment by oxtoacart** [comment]
@atavism Let me know when you're done QA'ing and I'll take a look.  I'm looking forward to reviewing it.

BTW - on the WebSocket reconnects, the use case that I'm thinking of is that the user opened the Lantern UI at some point, Lantern goes away but the old UI window stays open, and at some point in the future the user runs Lantern again.  It would be nice if that window started working again.  To that end, maybe we can just allow it to keep retrying?  As long as the interval isn't too small, it shouldn't be a big performance drain.

**Comment by atavism** [comment]
@oxtoacart Yeah, that use case is already supported. Currently, the websocket onclose event triggers an interval that tries to re-connect to Lantern every 5 seconds until a max number of reconnects has been reached. We could just keep trying to reconnect indefinitely though.

**Comment by oxtoacart** [comment]
> We could just keep trying to reconnect indefinitely though.

Yes, that's all I was asking for :)

**Comment by atavism** [comment]
> Yes, that's all I was asking for :)

Oh, cool :) Sure. I'll make that change.

File: src/github.com/getlantern/flashlight/http/http.go [hunk 1]
```diff
@@ -1,125 +1,59 @@
...
 
-type JsonResponse struct {
-	Error        string   `json:"Error, omitempty"`
-	ProxiedSites []string `json:"ProxiedSites, omitempty"`
-	Global       []string `json:"Global, omitempty"`
-}
-
-type ProxiedSitesHandler struct {
+type UIServer struct {
+	Conn             *websocket.Conn
```
Comment (oxtoacart) [inline src/github.com/getlantern/flashlight/http/http.go]: What if multiple clients connect?

Comment (atavism) [inline src/github.com/getlantern/flashlight/http/http.go]: I'm handling that case now. We could either manage a pool of connections or close the former whenever a new connection is open.

**Comment by atavism** [comment]
@oxtoacart @myleshorton The latest changes include support for handling multiple instances of the UI. Also, when you restart Lantern, you can leave off "-openui" and any existing UI window should automatically reactivate.

**Comment by oxtoacart** [comment]
@atavism Great!  Does this mean it's ready for review, or do you have anything left to do?

**Comment by atavism** [comment]
> @atavism Great! Does this mean it's ready for review, or do you have anything left to do?

I'm still doing some QAing and handling edge cases in the unit testing, but you can start reviewing the changes now if you'd like.

**Comment by oxtoacart** [comment]
That's ok, I'll wait till you're done. Just drop me a line when I have the green light.

**Comment by atavism** [comment]
> That's ok, I'll wait till you're done. Just drop me a line when I have the green light.

@oxtoacart You can go ahead and review this branch now. I'm done QAing the changes.

**Comment by oxtoacart** [comment]
> You can go ahead and review this branch now. I'm done QAing the changes.

Excellent, getting started now.

**Comment by oxtoacart** [comment]
LGTM.

**Pull request merged**